cachi2/core/package_managers/yarn_classic/main.py

```diff
@@ -125,3 +137,58 @@ def _verify_corepack_yarn_version(source_dir: RootedPath, env: dict[str, str]) -
         )
 
     log.info("Processing the request using yarn@%s", installed_yarn_version)
+
+
+def _verify_no_offline_mirror_collisions(packages: Iterable[YarnClassicPackage]) -> None:
+    """Verify that there are no duplicate tarballs in the offline mirror."""
+    tarballs = []
+    for p in packages:
+        if isinstance(p, (RegistryPackage, UrlPackage)):
+            tarball_name = _get_tarball_mirror_name(p.url)
+            tarballs.append(tarball_name)
+
+        elif isinstance(p, GitPackage):
+            tarball_name = _get_git_tarball_mirror_name(p.url)
+            tarballs.append(tarball_name)
+
+    c = Counter(tarballs)
+    duplicate_tarballs = [f"{name} ({count}x)" for name, count in c.most_common() if count > 1]
+    if len(duplicate_tarballs) > 0:
+        raise PackageManagerError(f"Duplicate tarballs detected: {', '.join(duplicate_tarballs)}")
+
+
+# https://github.com/yarnpkg/yarn/blob/7cafa512a777048ce0b666080a24e80aae3d66a9/src/fetchers/tarball-fetcher.js#L21
+RE_URL_NAME_MATCH = r"/(?:(@[^/]+)(?:\/|%2f))?[^/]+/(?:-|_attachments)/(?:@[^/]+\/)?([^/]+)$"
+
+
+# https://github.com/yarnpkg/yarn/blob/7cafa512a777048ce0b666080a24e80aae3d66a9/src/fetchers/tarball-fetcher.js#L65
+def _get_tarball_mirror_name(url: str) -> str:
+    parsed_url = urlparse(url)
+    path = Path(parsed_url.path)
+
+    match = re.search(RE_URL_NAME_MATCH, str(path))
+
+    if match:
```

Comment:
Non-blocking PEP8 zeal: match could be `None` so it is better to be explicit by checking if it `is` th` `None`.